lib/axon/training.ex

```diff
@@ -144,45 +147,49 @@ defmodule Axon.Training do
   ## Helpers
 
   defp train_epoch(step_fn, model_state, inputs, targets, epoch, opts) do
-    total_batches = Enum.count(inputs)
-
     {log_every, jit_opts} = Keyword.pop(opts, :log_every)
 
-    log_freq =
-      case log_every do
-        :epoch ->
-          total_batches
+    dataset =
+      inputs
+      |> Stream.zip(targets)
+      |> Stream.with_index()
 
-        x when is_integer(x) ->
-          x
+    {model_state, avg_loss, total_batches} =
+      for {{inp, tar}, i} <- dataset, reduce: {model_state, Nx.tensor(0.0), 0} do
```

Comment:
I'm guessing this runs to the end of the stream? I think perhaps we'll want an option to terminate before the end of a stream. I'm trying to think of the best way to do that without loading `n` steps all into memory at once with something like `Enum.take`, any thoughts?

Reply:
You can do a `Stream.take(n)` or do a throw/try+catch.

Reply:
Can we go with `Enum.reduce_while/3`? We just need the accumulator when processing each batch, but does that function keep the accumulator as well as the first _n_ elements of the stream in memory at once? With the continue/halting function of `reduce_while` it would easy to use incorporate different early stopping criteria.

Reply:
@seanmor5 @josevalim, do you want me to add options for early stopping in this PR? Otherwise it'll be a straightforward addition to convert the comprehension to `reduce_while` later on.

[Early stopping](https://pytorch-lightning.readthedocs.io/en/latest/common/early_stopping.html#stopping-an-epoch-early) might be specified by a function that accepts a combination of current batch/step loss, previous batch loss, average loss, and batch index and returns a boolean indicating whether training should continue.

Reply:
@t-rutten I would hold off on that for now, I would like to implement that as a callback when those are added into the training API

Reply:
Makes sense @seanmor5. Do you have any more suggestions for the changes here?

Reply:
No changes, I'm happy with where this is, unless you have anything else you'd like to add, I'll merge :)

Reply:
No changes, I'm happy with where this is, unless you have anything else you'd like to add, I'll merge :)